vecindex: create helper structs to search vector index

andy-kimball opened the pull request:
Add helper structs that can be embedded in processors and used by the execution engine to search the vector index. These encapsulate details so that the execution engine doesn't need to understand them.

Epic: CRDB-42943

Release note: None

andy-kimball committed:
vecindex: create helper structs to search vector index

Add helper structs that can be embedded in processors and used by the
execution engine to search the vector index. These encapsulate details
so that the execution engine doesn't need to understand them.

Epic: CRDB-42943

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142176)
<!-- Reviewable:end -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice

> Reviewed 14 of 14 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142176#-:-OKYNScU09PPmwWxs607:bb6m2cs)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/rowexec/vector_search.go` line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142176#-OKS7wL0ATPfw77zBhrr:-OKS7wL0ATPfw77zBhrs:b1955cn) ([raw file](https://github.com/cockroachdb/cockroach/blob/161b5891025a05e0ba97c698ee3b2718c4c5944a/pkg/sql/rowexec/vector_search.go#L302)):*
> ```Go
> 				}
> 				// It is possible for the search not to find the target index entry, in
> 				// which case the result is nil.
> ```

nit: this comment needs to be updated.
___
*[`pkg/sql/vecindex/mutation_searcher.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142176#-OKSC-jpABNz4nBnFelX:-OKSC-jpABNz4nBnFelY:b5qi5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/161b5891025a05e0ba97c698ee3b2718c4c5944a/pkg/sql/vecindex/mutation_searcher.go#L30)):*
> ```Go
> 	txn          vecstore.Txn
> 	idxCtx       cspann.Context
> 	partitionKey cspann.PartitionKey
> ```

nit: since `encoded` is a Datum, why not do the same for the partition key? With the requirement that `tree.DNull` be returned in place of the invalid partition.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142176)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/rowexec/vector_search.go` line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142176#-OKS7wL0ATPfw77zBhrr:-OKYPFRh8tuL8rJDJFBM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/161b5891025a05e0ba97c698ee3b2718c4c5944a/pkg/sql/rowexec/vector_search.go#L302)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: this comment needs to be updated.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/mutation_searcher.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142176#-OKSC-jpABNz4nBnFelX:-OKYT8tK120L0yluEE94:bkqbcol) ([raw file](https://github.com/cockroachdb/cockroach/blob/161b5891025a05e0ba97c698ee3b2718c4c5944a/pkg/sql/vecindex/mutation_searcher.go#L30)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: since `encoded` is a Datum, why not do the same for the partition key? With the requirement that `tree.DNull` be returned in place of the invalid partition.
</blockquote></details>

OK, I changed it, though I just used `nil` rather than `tree.DNull`.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
*[`pkg/sql/vecindex/mutation_searcher.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142176#-OKSC-jpABNz4nBnFelX:-OKYdoOMDTH4uCTmimYm:blmkv8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/161b5891025a05e0ba97c698ee3b2718c4c5944a/pkg/sql/vecindex/mutation_searcher.go#L30)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

OK, I changed it, though I just used `nil` rather than `tree.DNull`.
</blockquote></details>

Actually, I see now that the exec engine is expecting `tree.DNull` for this case, so I changed it to that.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r=drewkimball
